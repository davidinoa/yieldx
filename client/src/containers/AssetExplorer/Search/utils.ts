interface InGridFilter {
  filter: string | number;
  filterTo: number;
  filterType: string;
  type: string;
  dateFrom: string;
  dateTo: string;
  values: string[];
}

export interface AgGridFilters {
  [key: string]: any;
}

const getFilter = (filter: InGridFilter) => {
  /**
   *  The filter object from Ag Grid API. In the case of filtering `name` column.
   *  { filter: "Asset Name", filterType: "text", type: "contains"}
   */
  let searchFilter = {};
  const {
    filter: filterContent,
    filterType,
    type,
    filterTo,
    values,
    dateFrom,
    dateTo
  } = filter;
  if (filterType === "text") {
    switch (type) {
      case "contains":
        searchFilter = { containsText: filterContent };
        break;
      case "notContains":
        searchFilter = { doesNotContainText: filterContent };
        break;
      case "equals":
        searchFilter = { equalsText: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualText: filterContent };
        break;
      case "startsWith":
        searchFilter = { startsWithText: filterContent };
        break;
      case "endsWith":
        searchFilter = { endsWith: filterContent };
        break;
      default:
        searchFilter = { containsText: filterContent };
        break;
    }
  } else if (filterType === "number") {
    switch (type) {
      case "equals":
        searchFilter = { equalsNumber: filterContent };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualNumber: filterContent };
        break;
      case "lessThan":
        searchFilter = { isLessThanNumber: filterContent };
        break;
      case "lessThanOrEqual":
        searchFilter = { isLessThanOrEqualToNumber: filterContent };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanNumber: filterContent };
        break;
      case "greaterThanOrEqual":
        searchFilter = { isGreaterThanOrEqualToNumber: filterContent };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToNumber: filterContent,
          isLessThanOrEqualToNumber: filterTo
        };
        break;
      default:
        searchFilter = { equalsNumber: filterContent };
        break;
    }
  } else if (filterType === "date") {
    const convertedDateFrom = new Date(dateFrom);
    const convertedDateTo = new Date(dateTo);
    switch (type) {
      case "equals":
        searchFilter = { equalsDate: convertedDateFrom };
        break;
      case "greaterThan":
        searchFilter = { isGreaterThanDate: convertedDateFrom };
        break;
      case "lessThan":
        searchFilter = { isLessThanDate: convertedDateFrom };
        break;
      case "notEqual":
        searchFilter = { doesNotEqualDate: convertedDateFrom };
        break;
      case "inRange":
        searchFilter = {
          isGreaterThanOrEqualToDate: convertedDateFrom,
          isLessThanOrEqualToDate: convertedDateTo
        };
        break;
      default:
        searchFilter = { equalsDate: convertedDateFrom };
        break;
    }
  } else if (filterType === "set" && values) {
    // We may need to add `set` type Ag grid column filter params like cusip or isin
    // in the search request in the future.
    // searchFilter = { inText: values };
  }
  return searchFilter;
};

export const convertInGridFilters = (filters: AgGridFilters) => {
  const convertedFilters: { [index: string]: any } = {};
  Object.keys(filters).forEach(key => {
    const inGridFilter = filters[key];
    if (key === "minimumQuantity") key = "bond.minimumDenomination";
    convertedFilters[key] = getFilter(inGridFilter);
  });

  return convertedFilters;
};
