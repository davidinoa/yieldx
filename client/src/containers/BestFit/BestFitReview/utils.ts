import { getRatingValue } from "../../../util/ratings";
import YieldXPortfolio, {
  YieldXPosition
} from "../../../models/YieldXPortfolio";

export function getBondRatingsData(
  portfolio: YieldXPortfolio
): { name: string; value: number }[] {
  if (portfolio.getBondSleeve) {
    const ratingMarketValues: { [key: string]: number } = {};
    const bondPortfolio = portfolio.getBondSleeve();
    bondPortfolio.positions.forEach((position: YieldXPosition) => {
      const { asset, weight } = position;
      const rating = asset.rating ?? "NR";
      ratingMarketValues[rating] = (ratingMarketValues[rating] || 0) + weight;
    });
    return Object.entries(ratingMarketValues)
      .map(([rating, weight]) => {
        return {
          name: rating,
          value: weight
        };
      })
      .sort((a, b) => {
        return getRatingValue(a.name) - getRatingValue(b.name);
      });
  }
  return [];
}
