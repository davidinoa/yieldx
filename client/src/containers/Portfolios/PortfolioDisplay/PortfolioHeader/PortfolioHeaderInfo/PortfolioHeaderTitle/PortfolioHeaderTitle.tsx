import React from "react";
import { Row } from "react-bootstrap";
import { useHistory } from "react-router";
import { EditIcon, InfoName } from "../../PortfolioHeader.styles";

export type PortfolioDisplayStatus = "PAPER" | "LIVE" | "CLOSED" | "PENDING";

interface ReviewProps {
  portfolioName?: never;
  isReview: boolean;
  isInpaas: boolean;
  status?: PortfolioDisplayStatus;
}

interface SavedProps {
  portfolioName: string;
  isReview?: never;
  isInpaas?: never;
  status?: PortfolioDisplayStatus;
}

export default function PortfolioHeaderTitle({
  portfolioName,
  isReview,
  isInpaas,
  status
}: ReviewProps | SavedProps) {
  const history = useHistory();
  return (
    <Row>
      <InfoName>
        {isReview ? (
          <h1>Review your {isInpaas ? "proposal" : "portfolio"}</h1>
        ) : (
          <h1>
            {portfolioName}
            {status !== "CLOSED" && (
              <button
                type="button"
                data-testid="edit-name-button"
                // TODO: make edit name modal open with local state instead of url hash
                onClick={() =>
                  history.push(`${history.location.pathname}#edit-name`)
                }
              >
                <EditIcon />
              </button>
            )}
          </h1>
        )}
      </InfoName>
    </Row>
  );
}
