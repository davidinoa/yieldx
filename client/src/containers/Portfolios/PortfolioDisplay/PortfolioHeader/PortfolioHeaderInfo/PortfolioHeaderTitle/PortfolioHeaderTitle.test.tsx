import React from "react";
import { render, screen } from "../../../../../../test/utils";
import PortfolioHeaderTitle from "./PortfolioHeaderTitle";

describe("PortfolioHeaderTitle", () => {
  it("Should render properly if review for custom (best fit / asset explorer)", () => {
    const { asFragment } = render(<PortfolioHeaderTitle isInpaas={false} isReview />);
    expect(screen.queryByText(/review your portfolio/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it("Should render properly if review for inpaas", () => {
    const { asFragment } = render(<PortfolioHeaderTitle isInpaas isReview />);
    expect(screen.queryByText(/review your proposal/i)).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
  it("Should render portfolio name and edit button if not review", () => {
    const { asFragment } = render(<PortfolioHeaderTitle portfolioName="Mock portfolio" />);
    expect(screen.queryByText(/mock portfolio/i)).toBeInTheDocument();
    expect(screen.queryByTestId("edit-name-button")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
