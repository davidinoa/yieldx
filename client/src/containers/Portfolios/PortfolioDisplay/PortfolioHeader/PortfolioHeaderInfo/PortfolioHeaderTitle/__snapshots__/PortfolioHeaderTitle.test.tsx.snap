// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PortfolioHeaderTitle Should render portfolio name and edit button if not review 1`] = `
<DocumentFragment>
  .c0 h1 {
  line-height: 1.1em;
  margin: 0;
  font-size: 1.5rem;
}

.c0 h1 svg {
  margin-left: 0.5rem;
}

.c0 button {
  background: none;
  border: none;
  padding: 0;
  margin-left: 0.25rem;
}

@media (min-width:1600px) {
  .c0 h1 {
    font-size: 1.95rem;
  }
}

<div
    class="row"
  >
    <div
      class="c0 col-24"
    >
      <h1>
        Mock portfolio
        <button
          data-testid="edit-name-button"
          type="button"
        >
          <svg
            class=""
          >
            edit.svg
          </svg>
        </button>
      </h1>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PortfolioHeaderTitle Should render properly if review for custom (best fit / asset explorer) 1`] = `
<DocumentFragment>
  .c0 h1 {
  line-height: 1.1em;
  margin: 0;
  font-size: 1.5rem;
}

.c0 h1 svg {
  margin-left: 0.5rem;
}

.c0 button {
  background: none;
  border: none;
  padding: 0;
  margin-left: 0.25rem;
}

@media (min-width:1600px) {
  .c0 h1 {
    font-size: 1.95rem;
  }
}

<div
    class="row"
  >
    <div
      class="c0 col-24"
    >
      <h1>
        Review your portfolio
      </h1>
    </div>
  </div>
</DocumentFragment>
`;

exports[`PortfolioHeaderTitle Should render properly if review for inpaas 1`] = `
<DocumentFragment>
  .c0 h1 {
  line-height: 1.1em;
  margin: 0;
  font-size: 1.5rem;
}

.c0 h1 svg {
  margin-left: 0.5rem;
}

.c0 button {
  background: none;
  border: none;
  padding: 0;
  margin-left: 0.25rem;
}

@media (min-width:1600px) {
  .c0 h1 {
    font-size: 1.95rem;
  }
}

<div
    class="row"
  >
    <div
      class="c0 col-24"
    >
      <h1>
        Review your proposal
      </h1>
    </div>
  </div>
</DocumentFragment>
`;
